pkg/manager/member/shared_store_failover.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+)
+
+// StoreCommonAccess contains the set of functions to access the properties of TiKV and TiFlash types in a common way
+type StoreCommonAccess interface {
+	GetMemberType() v1alpha1.MemberType
+	GetMaxFailoverCount(tc *v1alpha1.TidbCluster) *int32
+	GetStores(tc *v1alpha1.TidbCluster) map[string]v1alpha1.TiKVStore
+	CreateFailureStoresIfAbsent(tc *v1alpha1.TidbCluster)
+	GetFailureStores(tc *v1alpha1.TidbCluster) map[string]v1alpha1.TiKVFailureStore
+	SetFailoverUIDIfAbsent(tc *v1alpha1.TidbCluster)
+	SetFailureStores(tc *v1alpha1.TidbCluster, storeID string, failureStore v1alpha1.TiKVFailureStore)
+	ClearFailStatus(tc *v1alpha1.TidbCluster)
+	GetStsDesiredOrdinals(tc *v1alpha1.TidbCluster, excludeFailover bool) sets.Int32
+}
+
+// sharedStoreFailover contains the shared failover logic of TiKV and TiFlash
+type sharedStoreFailover struct {
+	storeAccess StoreCommonAccess
+	deps        *controller.Dependencies
+}
+
+func (ssf *sharedStoreFailover) doFailover(tc *v1alpha1.TidbCluster, failoverPeriod time.Duration) error {
```

Comment:
Maybe just make rename it as `Failover` to make `sharedStoreFailover` an implementation of `Failover`.